[region-isolation] Do not look through begin_borrow or move_value if they are marked as a var_decl.

gottesmm opened the pull request:
Just chopping off parts of a larger piece of work while I work at the same time on another piece of work.

----

The reason why we do this is that we want to treat this as a separate value from
their operand since they are the result of defining a new value.

This has a few nice side-effects, one of which is that if a let results in just
a begin_borrow [var_decl], we emit names for it.

I also did a little work around helping variable name utils to lookup names from
applies that are fed into a {move_value,begin_borrow} [var_decl] which then has
the debug_value we are searching for.

gottesmm committed:
[region-isolation] Add a small helper to TrackableValue to print it's isolation info.

I am finding that I am calling that a bunch so this just makes it a little more
convenient.

gottesmm committed:
[region-isolation] Do not look through begin_borrow or move_value if they are marked as a var_decl.

The reason why we do this is that we want to treat this as a separate value from
their operand since they are the result of defining a new value.

This has a few nice side-effects, one of which is that if a let results in just
a begin_borrow [var_decl], we emit names for it.

I also did a little work around helping variable name utils to lookup names from
applies that are fed into a {move_value,begin_borrow} [var_decl] which then has
the debug_value we are searching for.

gottesmm committed:
[region-isolation] Move SILIsolationInfo determining code for ref_element_addr and global_addr onto SILIsolationInfo and call that instead.

gottesmm committed:
[region-isolation] Eliminate unnecessary using TrackableValueID = Element.

Having two artificial typedefs for the same wrapped value is just confusing.
Better to just have one and make the code simpler to understand.

gottesmm committed:
[region-isolation] Rename Partition::{fresh_label,freshLabel}.

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
[region-isolation] Move computation of SILIsolationInfo for FunctionRefInst/ClassMethodInst into SILIsolationInfo::get instead of handrolling in RegionAnalysis.

gottesmm committed:
[region-isolation] Change load [copy]/load_borrow to just use SILIsolationInfo::get instead of computing actor isolation by hand.

Just more recoring on top of SILIsolationInfo::get.

gottesmm committed:
[region-isolation] Finish moving isolation computation into SILIsolationInfo::get.

gottesmm committed:
[region-isolation] Make load/load_borrow look through instructions.

This ensures that when we process, we consider load/load_borrow's result to be
equivalent to its operand. This ensures that a load/load_borrow cannot act as a
use of its operand preventing spurious diagnostics.

gottesmm commented:
This is actually going to cause a merge conflict with a different PR I have going. I am going to do some more work here and rebase on top of that. That means I am going to put my SILIsolationInfo refactoring work here as well.

gottesmm committed:
[region-isolation] Begin tracking in SILIsolationInfo the actorInstance that a value is isolated to if we are dealing with an actor instance.

This will let us distinguish in between values derived from two actor instances
of the same type and to emit better errors.

gottesmm committed:
[variable-name-utils] Refactor inferName/inferNameAndRoot helpers onto VariableNameInferrer.

I have been using these in TransferNonSendable and they are useful in terms of
reducing the amount of code that one has to type to use this API. I am going to
need to use it in SILIsolationInfo, so it makes sense to move it into
SILOptimizer/Utils.

NFCI.

gottesmm committed:
[region-isolation] When printing a SILIsolationInfo description for diagnostics, if we have a SIL actor instance, print -isolated instead of actor-isolated.

rdar://122501400

gottesmm committed:
[region-isolation] Emit the correct error for closures that capture actor self.

rdar://122501400

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

ktoso commented on include/swift/SILOptimizer/Utils/PartitionUtils.h:
```diff
@@ -118,14 +118,24 @@ class SILIsolationInfo {
+        isolatedValue(isolatedValue), actorInstance(actorInstance) {
+    assert((!actorInstance ||
+            (actorIsolation.getKind() == ActorIsolation::ActorInstance &&
+             actorInstance->getType().isActor())) &&
```
```suggestion
             actorInstance->getType().isAnyActor())) &&
```

otherwise would crash on a distributed actor

gottesmm replied:
This actually calls:

```
bool TypeBase::isActorType() {                                                                                            
  if (auto actor = getAnyActor())                                                                                         
    return actor->isActor();                                                                                              
  return false;                                                                                                           
} 
```

which does any actor.

ktoso replied:
Right, which is still wrong -- it does isActor but should do `isAnyActor` here

ktoso commented:
Not much I can say about the analysis bits but looked good :)

gottesmm committed:
[region-isolation] Do not treat functions/class_methods that are isolated to a #isolated as being globally isolated to that.

Instead, we need to consider the isolation at the apply site.

rdar://126285681
rdar://125078448

gottesmm committed:
[region-isolation] Fix thinko.

The thinko was that rather than use the actual isolated parameter of a partial apply as the isolated parameter, I instead just used the last operand... :doh:. This fixes:

rdar://126297097

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Lets try this again

gottesmm committed:
Remove pragma clang optimize off that snuck in.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

ktoso reviewed (dismissed):
Sorry but I still think this assertion will wrongly crash for distributed actor isolated code -- commented inline

gottesmm committed:
[region-isolation] Add some tests around initializers and distributed actors.

gottesmm committed:
[region-isolation] Out of an abundance of caution convert isActor -> isAnyActor().

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
I also included some tests around distributed actors as well to validate that we do not crash.
